In [5]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import math
%matplotlib inline

In [6]:
df = pd.read_csv('bank-additional-full.csv', sep=';')

In [7]:
# There are 101,766 rows and 50 columns
print(df.shape)

(41188, 21)


In [9]:
# Examine data, there are some hidden columns
# show all columns
df.columns

Index(['age', 'job', 'marital', 'education', 'default', 'housing', 'loan',
       'contact', 'month', 'day_of_week', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'emp.var.rate', 'cons.price.idx',
       'cons.conf.idx', 'euribor3m', 'nr.employed', 'y'],
      dtype='object')

In [10]:
# examine the common stats of numeric columns
print(df.describe())

               age      duration      campaign         pdays      previous  \
count  41188.00000  41188.000000  41188.000000  41188.000000  41188.000000   
mean      40.02406    258.285010      2.567593    962.475454      0.172963   
std       10.42125    259.279249      2.770014    186.910907      0.494901   
min       17.00000      0.000000      1.000000      0.000000      0.000000   
25%       32.00000    102.000000      1.000000    999.000000      0.000000   
50%       38.00000    180.000000      2.000000    999.000000      0.000000   
75%       47.00000    319.000000      3.000000    999.000000      0.000000   
max       98.00000   4918.000000     56.000000    999.000000      7.000000   

       emp.var.rate  cons.price.idx  cons.conf.idx     euribor3m   nr.employed  
count  41188.000000    41188.000000   41188.000000  41188.000000  41188.000000  
mean       0.081886       93.575664     -40.502600      3.621291   5167.035911  
std        1.570960        0.578840       4.628198    

In [12]:
# loop through the data of each column, find the column has value of cell which is object =='?' then count

for col in df.columns:
    if df[col].dtype == object:
         print(col, df[col][df[col] == '?'].count())

# There are 101,766 rows and 50 columns
print(df.shape)

job 0
marital 0
education 0
default 0
housing 0
loan 0
contact 0
month 0
day_of_week 0
poutcome 0
y 0
(41188, 21)


In [8]:
df.head()

age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0   may         mon  ...         1    999         0  nonexistent          1.1   
1   may         mon  ...         1    999         0  nonexistent          1.1   
2   may         mon  ...         1    999         0  nonexistent          1.1   
3   may         mon  ...         1    999         0  nonexistent          1.1   
4   may         mon  ...         1    999         0  nonexistent          1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0          93.994          -36.4      4.857       5191.0  no  
1          93.994          -36.4      4.857       5191.0  no  
2          93.994          -36.4      4.857       5191.0  no  
3          93.994          -36.4      4.857       5191.0  no  
4          93.994          -36.4      4.857       5191.0  no  

[5 rows x 21 columns]

In [14]:
# Because there are many of hidden columns & hidden rows,
#in order to have an overview of dataframe, try to see through all of them
#there are a lot missing value columns & cells
pd.set_option('display.max_columns', None)
#pd.set_option('display.max_rows', None)
df

age            job   marital            education  default  housing  \
0       56      housemaid   married             basic.4y       no       no   
1       57       services   married          high.school  unknown       no   
2       37       services   married          high.school       no      yes   
3       40         admin.   married             basic.6y       no       no   
4       56       services   married          high.school       no       no   
5       45       services   married             basic.9y  unknown       no   
6       59         admin.   married  professional.course       no       no   
7       41    blue-collar   married              unknown  unknown       no   
8       24     technician    single  professional.course       no      yes   
9       25       services    single          high.school       no      yes   
10      41    blue-collar   married              unknown  unknown       no   
11      25       services    single          high.school       no      yes   
12      29    blue-collar    single          high.school       no       no   
13      57      housemaid  divorced             basic.4y       no      yes   
14      35    blue-collar   married             basic.6y       no      yes   
15      54        retired   married             basic.9y  unknown      yes   
16      35    blue-collar   married             basic.6y       no      yes   
17      46    blue-collar   married             basic.6y  unknown      yes   
18      50    blue-collar   married             basic.9y       no      yes   
19      39     management    single             basic.9y  unknown       no   
20      30     unemployed   married          high.school       no       no   
21      55    blue-collar   married             basic.4y  unknown      yes   
22      55        retired    single          high.school       no      yes   
23      41     technician    single          high.school       no      yes   
24      37         admin.   married          high.school       no      yes   
25      35     technician   married    university.degree       no       no   
26      59     technician   married              unknown       no      yes   
27      39  self-employed   married             basic.9y  unknown       no   
28      54     technician    single    university.degree  unknown       no   
29      55        unknown   married    university.degree  unknown  unknown   
...    ...            ...       ...                  ...      ...      ...   
41158   35     technician  divorced             basic.4y       no       no   
41159   35     technician  divorced             basic.4y       no      yes   
41160   33         admin.   married    university.degree       no       no   
41161   33         admin.   married    university.degree       no      yes   
41162   60    blue-collar   married             basic.4y       no      yes   
41163   35     technician  divorced             basic.4y       no      yes   
41164   54         admin.   married  professional.course       no       no   
41165   38      housemaid  divorced    university.degree       no       no   
41166   32         admin.   married    university.degree       no       no   
41167   32         admin.   married    university.degree       no      yes   
41168   38   entrepreneur   married    university.degree       no       no   
41169   62       services   married          high.school       no      yes   
41170   40     management  divorced    university.degree       no      yes   
41171   33        student   married  professional.course       no      yes   
41172   31         admin.    single    university.degree       no      yes   
41173   62        retired   married    university.degree       no      yes   
41174   62        retired   married    university.degree       no      yes   
41175   34        student    single              unknown       no      yes   
41176   38      housemaid  divorced          high.school       no      yes   
41177   57        retired   m

In [43]:
df.groupby(['poutcome', 'y']).mean()

age    duration  campaign       pdays  previous  \
poutcome    y                                                            
failure     no   39.626817  211.764464  2.028517  980.291198  1.139841   
            yes  41.871074  467.224793  1.864463  879.682645  1.342149   
nonexistent no   39.935167  221.565850  2.710567  999.000000  0.000000   
            yes  40.173193  623.688634  2.184973  999.000000  0.000000   
success     no   40.453027  241.175365  1.991649    5.624217  1.421712   
            yes  42.864653  363.680089  1.710291    5.568233  1.648770   

                 emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
poutcome    y                                                             
failure     no      -1.611489       93.052742     -43.337209   1.740383   
            yes     -2.131240       93.158255     -39.002645   1.112688   
nonexistent no       0.491993       93.672219     -40.300904   4.083050   
            yes     -0.808946       93.372055     -40.506304   2.669357   
success     no      -2.042589       93.165113     -39.477662   1.199499   
            yes     -2.117338       93.425035     -37.805034   0.887831   

                 nr.employed  
poutcome    y                 
failure     no   5099.794132  
            yes  5049.159504  
nonexistent no   5186.565289  
            yes  5125.858007  
success     no   5053.795198  
            yes  5018.206600

In [15]:
df.groupby(['job', 'y']).mean()

age    duration  campaign       pdays  previous  \
job           y                                                            
admin.        no   38.219846  215.093385  2.707718  981.270783  0.141125   
              yes  37.968935  517.414201  2.058432  773.512574  0.510355   
blue-collar   no   39.582057  229.878018  2.588208  991.516365  0.111653   
              yes  39.200627  732.673981  2.156740  899.324451  0.269592   
entrepreneur  no   41.703453  225.632883  2.579580  987.813814  0.128378   
              yes  41.935484  667.540323  2.064516  910.943548  0.250000   
housemaid     no   44.705451  218.778826  2.722222  986.491614  0.089099   
              yes  52.650943  535.537736  1.896226  727.367925  0.575472   
management    no   42.309707  220.828197  2.532357  984.848613  0.140986   
              yes  42.783537  543.804878  2.030488  786.929878  0.533537   
retired       no   59.926128  224.263608  2.657076  960.369362  0.212286   
              yes  68.253456  420.235023  1.942396  712.937788  0.668203   
self-employed no   40.176887  222.220912  2.724843  988.073113  0.123428   
              yes  38.006711  622.020134  2.114094  878.859060  0.315436   
services      no   38.090236  225.047998  2.619309  991.091607  0.136314   
              yes  36.077399  634.851393  2.232198  854.479876  0.365325   
student       no   26.396667  228.851667  2.238333  924.596667  0.395000   
              yes  24.800000  403.316364  1.810909  656.116364  0.807273   
technician    no   38.600033  213.452021  2.641776  985.135540  0.111093   
              yes  37.746575  553.190411  2.046575  793.676712  0.505479   
unemployed    no   39.844828  212.706897  2.668966  973.879310  0.131034   
              yes  39.062500  471.451389  1.930556  702.333333  0.611111   
unknown       no   45.375427  204.559727  2.679181  954.873720  0.095563   
              yes  47.054054  517.756757  2.405405  810.864865  0.621622   

                   emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
job           y                                                             
admin.        no       0.224068       93.566030     -40.349768   3.786567   
              yes     -1.383210       93.319537     -39.545488   1.965078   
blue-collar   no       0.303552       93.668887     -41.346565   3.835402   
              yes     -0.487774       93.491480     -41.770846   2.915715   
entrepreneur  no       0.242868       93.626245     -41.319820   3.881632   
              yes     -0.745161       93.381153     -40.895161   2.818839   
housemaid     no       0.641300       93.712759     -39.626834   4.243082   
              yes     -1.437736       93.350934     -38.311321   1.908717   
management    no       0.141217       93.550062     -40.619992   3.785477   
              yes     -1.230793       93.306631     -39.456402   2.232896   
retired       no      -0.283904       93.493273     -38.920762   3.249794   
              yes     -1.926267       93.245629     -37.542857   1.348569   
self-employed no       0.242610       93.591786     -40.501022   3.854206   
              yes     -1.173154       93.288470     -40.377852   2.282242   
services      no       0.256720       93.650433     -41.327537   3.793805   
              yes     -0.743034       93.456598     -40.866873   2.631139   
student       no      -1.133167       93.350790     -40.751500   2.208683   
              yes     -2.007636       93.289771     -38.957091   1.176313   
technician    no       0.441693       93.584196     -39.918310   4.012318   
              yes     -1.102055       93.374286     -40.003836   2.239584   
unemployed    no       0.147241       93.604369     -40.193908   3.775461   
              yes     -1.676389       93.318563     -38.881944   1.600444   
unknown       no       0.555290       93.748703     -38.882594   4.154038   
              yes     -1.205405       93.483270     -38.127027   2.325622   

                   nr.employed  
job         

In [ ]:
for i in df['y']:
    if i == no

Input variables:
# bank client data:
1 - age (numeric)

2 - job : type of job (categorical: 'admin.','blue-collar','entrepreneur','housemaid','management','retired','self-employed','services','student','technician','unemployed','unknown')

3 - marital : marital status (categorical: 'divorced','married','single','unknown'; note: 'divorced' means divorced or widowed)

4 - education (categorical: 'basic.4y','basic.6y','basic.9y','high.school','illiterate','professional.course','university.degree','unknown')

5 - default: has credit in default? (categorical: 'no','yes','unknown')

6 - housing: has housing loan? (categorical: 'no','yes','unknown')

7 - loan: has personal loan? (categorical: 'no','yes','unknown')
# related with the last contact of the current campaign:

8 - contact: contact communication type (categorical: 'cellular','telephone') 

9 - month: last contact month of year (categorical: 'jan', 'feb', 'mar', ..., 'nov', 'dec')

10 - day_of_week: last contact day of the week (categorical: 'mon','tue','wed','thu','fri')

11 - duration: last contact duration, in seconds (numeric). Important note: this attribute highly affects the output target (e.g., if duration=0 then y='no'). Yet, the duration is not known before a call is performed. Also, after the end of the call y is obviously known. Thus, this input should only be included for benchmark purposes and should be discarded if the intention is to have a realistic predictive model.

# other attributes:

12 - campaign: number of contacts performed during this campaign and for this client (numeric, includes last contact)

13 - pdays: number of days that passed by after the client was last contacted from a previous campaign (numeric; 999 means client was not previously contacted)

14 - previous: number of contacts performed before this campaign and for this client (numeric)

15 - poutcome: outcome of the previous marketing campaign (categorical: 'failure','nonexistent','success')

# social and economic context attributes

16 - emp.var.rate: employment variation rate - quarterly indicator (numeric)

17 - cons.price.idx: consumer price index - monthly indicator (numeric) 

18 - cons.conf.idx: consumer confidence index - monthly indicator (numeric) 

19 - euribor3m: euribor 3 month rate - daily indicator (numeric)

20 - nr.employed: number of employees - quarterly indicator (numeric)

Output variable (desired target):
21 - y - has the client subscribed a term deposit? (binary: 'yes','no')